src/core/ht.c

```diff
@@ -135,136 +148,147 @@ void get_imu_data(int bCalcDiff)
+    res = timer_settime(timerId, 0, &its, NULL);
+    if (res != 0){
+        LOGE("Error timer_settime: %s\n", strerror(errno));
+    }
 }
 
-static void iir_filter(float beta,float in, float* out)
+void set_maxangle_ht(int angle)
 {
-    *out = in*beta + (*out)*(1 - beta);
-}
-
-// Normalizes any number to an arbitrary range
-// by assuming the range wraps around when going below min or above max
-float normalize(float value, float start, float end)
-{
-  float width = end - start;          //
-  float offsetValue = value - start;  // value relative to 0
-
-  return (offsetValue - (floor(offsetValue / width) * width)) + start;
-  // + start to reset back to start of original range
+    ht_data.tiltFactor = 1000.0 / angle;
+    ht_data.rollFactor = 1000.0 / angle;
+    ht_data.panFactor = 1000.0 / angle;
 }
 
 // Rotate, in Order X -> Y -> Z
-void rotate(float pn[3], const float rotation[3])
+static void rotate(float pn[3], const float rot[3])
 {
-  float rot[3] = {0, 0, 0};
-  memcpy(rot, rotation, sizeof(rot[0]) * 3);
-
-  // Passed in Degrees
-  rot[0] *= DEG_TO_RAD;
-  rot[1] *= DEG_TO_RAD;
-  rot[2] *= DEG_TO_RAD;
-
   float out[3];
 
-  // X Rotation
-  if (rotation[0] != 0) {
+  // X-axis Rotation
+  if (rot[0] != 0) {
     out[0] = pn[0] * 1 + pn[1] * 0 + pn[2] * 0;
     out[1] = pn[0] * 0 + pn[1] * cos(rot[0]) - pn[2] * sin(rot[0]);
     out[2] = pn[0] * 0 + pn[1] * sin(rot[0]) + pn[2] * cos(rot[0]);
     memcpy(pn, out, sizeof(out[0]) * 3);
   }
 
-  // Y Rotation
-  if (rotation[1] != 0) {
+  // Y-axis Rotation
+  if (rot[1] != 0) {
     out[0] = pn[0] * cos(rot[1]) - pn[1] * 0 + pn[2] * sin(rot[1]);
     out[1] = pn[0] * 0 + pn[1] * 1 + pn[2] * 0;
     out[2] = -pn[0] * sin(rot[1]) + pn[1] * 0 + pn[2] * cos(rot[1]);
     memcpy(pn, out, sizeof(out[0]) * 3);
   }
 
-  // Z Rotation
-  if (rotation[2] != 0) {
+  // Z-axis Rotation
+  if (rot[2] != 0) {
     out[0] = pn[0] * cos(rot[2]) - pn[1] * sin(rot[2]) + pn[2] * 0;
     out[1] = pn[0] * sin(rot[2]) + pn[1] * cos(rot[2]) + pn[2] * 0;
     out[2] = pn[0] * 0 + pn[1] * 0 + pn[2] * 1;
     memcpy(pn, out, sizeof(out[0]) * 3);
   }
 }
 
-void calibrate_ht()
+static void calc_gyr(float* gyrAngle) //in degree
 {
-    uint16_t i;
-    float accAngle[3];
+    // convert gyro readings to degrees/sec (with calibration offsets)
+    gyrAngle[0] = gyr_to_dps(ht_data.sensor_data.gyr.x - ht_data.gyr_offset[0]);
+    gyrAngle[1] = gyr_to_dps(ht_data.sensor_data.gyr.y - ht_data.gyr_offset[1]);
+    gyrAngle[2] = gyr_to_dps(ht_data.sensor_data.gyr.z - ht_data.gyr_offset[2]);
+    rotate(gyrAngle, imu_orientation);
+}
+
+static void calc_acc(float* accAngle) //in G
+{
+    // convert accelerometer readings to G forces
+    accAngle[0] = acc_to_g(ht_data.sensor_data.acc.x);
+    accAngle[1] = acc_to_g(ht_data.sensor_data.acc.y);
+    accAngle[2] = acc_to_g(ht_data.sensor_data.acc.z);
+    rotate(accAngle, imu_orientation);
+}
 
+static int constrain(int value, int min, int max)
```

Comment:
shall we move these to a dedicated subdir eg `util`?

Reply:
So you want me to create a `util` folder and have a `math.c` & `math.h` file?
We could mode `normalize`, `constrain` to inline functions in the .h and move `rotate` to the .c file.

Reply:
yeah and probably `file_compare` and `file_exists` into something like `file.{c,h}`
for constrain a macro maybe more suitable tho, cover all the types.
i can submit a PR on top of this one, once its merged.

Reply:
Yeah, maybe a seperate PR for extracting utility functions and macros might be better

Reply:
I caved and did the refactoring to extract those functions 😃

Reply:
happens to the best of us :)